File: src/logic/airplay.ts
import { type AbstractDevice, type AirPlayClient, type AirPlayPlayer, Proto } from '@basmilius/apple-sdk';
import { type Device, Shortcuts } from '@basmilius/homey-common';
import type { AppleApp } from '../types';
import Homey from 'homey';
import AppleTVDevice from '../apple-tv/device';
import HomePodBaseDevice from '../homepod-base/device';
import { getFallbackArtworkUrl, repeatModeToCapability, safeCapabilityValue } from '../utils';

export type MiniPlayerState = {
    readonly deviceId: string;
    readonly deviceName: string;
    readonly track: string | null;
    readonly artist: string | null;
    readonly album: string | null;
    readonly playing: boolean | null;
    readonly position: number | null;
    readonly duration: number | null;
    readonly volume: number | null;
    readonly artworkUrl: string | null;
    readonly onoff: boolean | null;
    readonly shuffle: boolean;
    readonly repeat: string;
    readonly positionTimestamp: number;
    readonly features: {
        readonly previous: boolean;
        readonly next: boolean;
        readonly shuffle: boolean;
        readonly repeat: boolean;
    };
};

export default class AirPlayLogic extends Shortcuts<AppleApp> {
    get deviceName(): string {
        return this.#device.name;
    }

    get position(): number {
        return this.#sdkDevice?.state.elapsedTime ?? 0;
    }

    get positionTimestamp(): number {
        return Date.now();
    }

    get features(): { previous: boolean; next: boolean; shuffle: boolean; repeat: boolean } {
        return this.#getFeatureAvailability();
    }

    get repeat(): string {
        const repeatMode = this.#sdkDevice?.state.repeatMode;

        if (repeatMode === undefined) {
            return 'off';
        }

        switch (repeatMode) {
            case Proto.RepeatMode_Enum.One:
                return 'one';
            case Proto.RepeatMode_Enum.All:
                return 'all';
            default:
                return 'off';
        }
    }

    get shuffle(): boolean {
        const shuffleMode = this.#sdkDevice?.state.shuffleMode;

        if (shuffleMode === undefined) {
            return false;
        }

        return shuffleMode !== Proto.ShuffleMode_Enum.Off
            && shuffleMode !== Proto.ShuffleMode_Enum.Unknown;
    }

    readonly #device: Device<AppleApp, any>;

    #artwork!: Homey.Image;
    #artworkIdentifier?: string;
    #miniPlayerUpdateTimer?: NodeJS.Timeout;
    #nowPlayingDebounceTimer?: NodeJS.Timeout;
    #nowPlayingAppTimer?: NodeJS.Timeout;
    #sdkDevice?: AbstractDevice;
    #updateLock: Promise<void> = Promise.resolve();
    #volumeDebounceTimer?: NodeJS.Timeout;

    constructor(device: Device<AppleApp, any>) {
        super(device.app);

        this.#device = device;
        this.onNowPlayingChanged = this.onNowPlayingChanged.bind(this);
        this.onPlaybackStateChanged = this.onPlaybackStateChanged.bind(this);
        this.onVolumeChanged = this.onVolumeChanged.bind(this);
        this.onVolumeMutedChanged = this.onVolumeMutedChanged.bind(this);
        this.onArtworkChanged = this.onArtworkChanged.bind(this);
    }

    async initialize(): Promise<void> {
        await this.clearNowPlaying();

        this.#artwork = await this.#device.homey.images.createImage();
        await this.#device.setAlbumArtImage(this.#artwork);

        await this.updateArtworkUrl();
    }

    async uninitialize(): Promise<void> {
        clearTimeout(this.#miniPlayerUpdateTimer);
        clearTimeout(this.#nowPlayingDebounceTimer);
        clearTimeout(this.#nowPlayingAppTimer);
        clearTimeout(this.#volumeDebounceTimer);
        this.#removeListeners();

        try {
            await this.#artwork.unregister();
        } catch (err) {
            this.log(this.deviceName, 'Failed to unregister artwork image:', err);
        }
    }

    async clearNowPlaying(): Promise<void> {
        await this.#serialized(() => this.#clearNowPlayingImpl());
    }

    async #clearNowPlayingImpl(): Promise<void> {
        try {
            if (this.#artwork) {
                this.#artworkIdentifier = undefined;
                await this.#updateArtwork(null);
            }

            clearTimeout(this.#nowPlayingAppTimer);
            await this.#updateNowPlayingAppImpl(null, null);

            await Promise.allSettled([
                this.#device.setCapabilityValue('speaker_album', ''),
                this.#device.setCapabilityValue('speaker_artist', ''),
                this.#device.setCapabilityValue('speaker_track', ''),
                this.#device.setCapabilityValue('speaker_duration', -1),
                this.#device.setCapabilityValue('speaker_position', -1),
                this.#device.setCapabilityValue('speaker_playing', false),
                this.#device.hasCapability('speaker_repeat') ? this.#device.setCapabilityValue('speaker_repeat', 'none') : undefined,
                this.#device.hasCapability('speaker_shuffle') ? this.#device.setCapabilityValue('speaker_shuffle', false) : undefined
            ]);
            this.#emitMiniPlayerUpdate();
        } catch (err) {
            this.log(this.deviceName, 'Failed to clear now playing info', err);
        }
    }

    async updateArtworkUrl(): Promise<void> {
        if (!this.#device.hasCapability('artwork_url')) {
            return;
        }

        // @ts-expect-error The type definition for Homey.Image.cloudUrl does not exist, but the property is there.
        const cloudUrl = this.#artwork.cloudUrl;

        if (!cloudUrl) {
            return;
        }

        const cacheBuster = new Date().getTime();
        const artworkUrl = `${cloudUrl}?v=${cacheBuster}`;
        await this.#device.setCapabilityValue('artwork_url', artworkUrl);

        // @ts-expect-error The type definition for Homey.Image.localUrl does not exist, but the property is there.
        const localUrl = this.#artwork.localUrl;
        const localUrlWithCacheBuster = localUrl ? `${localUrl}?v=${cacheBuster}` : '';

        if (this.#device.hasCapability('artwork_url_cloud')) {
            await this.#device.setCapabilityValue('artwork_url_cloud', artworkUrl);
        }

        if (this.#device.hasCapability('artwork_url_local')) {
            await this.#device.setCapabilityValue('artwork_url_local', localUrlWithCacheBuster);
        }

        if (this.#device instanceof AppleTVDevice) {
            await this.app.appleTvFlow.triggerArtworkUrlUpdated(this.#device, localUrlWithCacheBuster, artworkUrl);
        } else if (this.#device instanceof HomePodBaseDevice) {
            await this.app.homePodFlow.triggerArtworkUrlUpdated(this.#device, localUrlWithCacheBuster, artworkUrl);
        }

        this.#emitMiniPlayerUpdate();
    }

    getState(): MiniPlayerState {
        return {
            deviceId: this.#device.id,
            deviceName: this.#device.getName(),
            track: safeCapabilityValue(this.#device, 'speaker_track'),
            artist: safeCapabilityValue(this.#device, 'speaker_artist'),
            album: safeCapabilityValue(this.#device, 'speaker_album'),
            playing: safeCapabilityValue(this.#device, 'speaker_playing'),
            position: this.position,
            duration: safeCapabilityValue(this.#device, 'speaker_duration'),
            volume: safeCapabilityValue(this.#device, 'volume_set'),
            artworkUrl: safeCapabilityValue(this.#device, 'artwork_url'),
            onoff: safeCapabilityValue(this.#device, 'onoff'),
            shuffle: this.shuffle,
            repeat: this.repeat,
            positionTimestamp: Date.now(),
            features: this.#getFeatureAvailability()
        };
    }

    emitUpdate(): void {
        this.#emitMiniPlayerUpdate();
    }

    setDevice(sdkDevice: AbstractDevice): void {
        this.#removeListeners();

        this.#sdkDevice = sdkDevice;
        sdkDevice.state.on('nowPlayingChanged', this.onNowPlayingChanged);
        sdkDevice.state.on('playbackStateChanged', this.onPlaybackStateChanged);
        sdkDevice.state.on('volumeChanged', this.onVolumeChanged);
        sdkDevice.state.on('volumeMutedChanged', this.onVolumeMutedChanged);
        sdkDevice.state.on('artworkChanged', this.onArtworkChanged);
    }

    #removeListeners(): void {
        if (!this.#sdkDevice) {
            return;
        }

        this.#sdkDevice.state.off('nowPlayingChanged', this.onNowPlayingChanged);
        this.#sdkDevice.state.off('playbackStateChanged', this.onPlaybackStateChanged);
        this.#sdkDevice.state.off('volumeChanged', this.onVolumeChanged);
        this.#sdkDevice.state.off('volumeMutedChanged', this.onVolumeMutedChanged);
        this.#sdkDevice.state.off('artworkChanged', this.onArtworkChanged);
    }

    async onNowPlayingChanged(client: AirPlayClient | null, _player: AirPlayPlayer | null): Promise<void> {
        clearTimeout(this.#nowPlayingDebounceTimer);
        this.#nowPlayingDebounceTimer = setTimeout(() => {
            this.#serialized(async () => {
                if (!client) {
                    await this.#clearNowPlayingImpl();
                    return;
                }

                await this.#updateNowPlaying();
            }).catch(err => this.log(this.deviceName, 'Failed to process now playing change:', err));
        }, 300);
    }

    async onPlaybackStateChanged(_client: AirPlayClient, _player: AirPlayPlayer, _oldState: Proto.PlaybackState_Enum, newState: Proto.PlaybackState_Enum): Promise<void> {
        // Fast path: update speaker_playing immediately without debounce.
        // The full nowPlayingChanged event will handle the rest.
        try {
            const isPlaying = newState === Proto.PlaybackState_Enum.Playing;
            await this.#device.setCapabilityValue('speaker_playing', isPlaying);
            this.#emitMiniPlayerUpdate();
        } catch (err) {
            this.log(this.deviceName, 'Failed to update playback state', err);
        }
    }

    async onArtworkChanged(_client: AirPlayClient, _player: AirPlayPlayer): Promise<void> {
        const client = this.#sdkDevice?.state.activeClient;

        if (!client) {
            return;
        }

        // Reset the artwork identifier so that #setArtwork re-fetches even
        // if the artworkId hasn't changed (the artwork DATA has changed).
        this.#artworkIdentifier = undefined;

        await this.#serialized(async () => {
            await this.#setArtwork(client);
            this.#emitMiniPlayerUpdate();
        }).catch(err => this.log(this.deviceName, 'Failed to process artwork change:', err));
    }

    onVolumeMutedChanged(muted: boolean): void {
        if (!this.#device.hasCapability('volume_mute')) {
            return;
        }

        this.#device.setCapabilityValue('volume_mute', muted)
            .catch(err => this.log(this.deviceName, 'Failed to update volume mute', err));
    }

    onVolumeChanged(): void {
        clearTimeout(this.#volumeDebounceTimer);
        this.#volumeDebounceTimer = setTimeout(() => {
            this.#updateVolume()
                .catch(err => this.log(this.deviceName, 'Failed to update volume:', err));
        }, 300);
    }

    async #updateVolume(): Promise<void> {
        if (!this.#device.hasCapability('volume_set')) {
            return;
        }

        const volume = this.#sdkDevice?.state.volume ?? 0;
        await this.#device.setCapabilityValue('volume_set', volume);
        this.#emitMiniPlayerUpdate();
    }

    async #setArtwork(client: AirPlayClient): Promise<void> {
        const artworkId = client.artworkId;

        if (artworkId === this.#artworkIdentifier) {
            return;
        }

        if (!artworkId) {
            this.#artworkIdentifier = undefined;
            await this.#updateArtwork(getFallbackArtworkUrl(client.bundleIdentifier));
            return;
        }

        try {
            this.#artworkIdentifier = artworkId;
            const artwork = await this.#sdkDevice?.artwork.get(600);

            if (artwork?.url) {
                await this.#updateArtwork(artwork.url);
            } else if (artwork?.data) {
                await this.#updateArtworkBuffer(artwork.data);
            } else {
                await this.#updateArtwork(null);
            }
        } catch (err) {
            this.#device.error(this.deviceName, 'Failed to fetch artwork', err);
        }
    }

    async #updateArtwork(url: string | null): Promise<void> {
        try {
            if (url) {
                this.#artwork.setUrl(url.replace('.heic', '.jpg'));
            } else {
                // @ts-expect-error The type definition of Homey.Image.setUrl() is incorrect.
                this.#artwork.setUrl(null);
            }

            await this.#artwork.update();
            await this.updateArtworkUrl();
        } catch (err) {
            this.log(this.deviceName, 'Failed to update album artwork', err);
        }
    }

    async #updateArtworkBuffer(buffer: Uint8Array<ArrayBufferLike> | Buffer): Promise<void> {
        const imageBuffer = Buffer.from(buffer);

        if (this.#isHeicBuffer(imageBuffer)) {
            return;
        }

        this.#artwork.setStream((stream: NodeJS.WritableStream) => {
            stream.end(imageBuffer);
        });
        await this.#artwork.update();
        await this.updateArtworkUrl();
    }

    #isHeicBuffer(buffer: Buffer): boolean {
        if (buffer.byteLength < 12) {
            return false;
        }

        const ftyp = buffer.subarray(4, 8).toString('ascii');

        if (ftyp !== 'ftyp') {
            return false;
        }

        const brand = buffer.subarray(8, 12).toString('ascii');
        return ['heic', 'heix', 'hevc', 'heim', 'heis', 'mif1'].includes(brand);
    }

    async #updateNowPlaying(): Promise<void> {
        if (this.#device.hasCapability('onoff') && this.#device.getCapabilityValue('onoff') === false) {
            return;
        }

        const client = this.#sdkDevice?.state.activeClient;

        if (!client) {
            return;
        }

        try {
            await this.#syncDynamicCapabilities(client);

            await Promise.allSettled([
                this.#device.setCapabilityValue('speaker_playing', client.isPlaying),
                this.#device.setCapabilityValue('speaker_album', client.album),
                this.#device.setCapabilityValue('speaker_artist', client.artist || client.activePlayer?.currentItemMetadata?.trackArtistName || client.displayName || '-'),
                this.#device.setCapabilityValue('speaker_track', client.title),
                this.#device.setCapabilityValue('speaker_duration', client.duration),
                this.#device.setCapabilityValue('speaker_position', client.elapsedTime)
            ]);

            if (this.#device.hasCapability('speaker_repeat')) {
                await this.#device.setCapabilityValue('speaker_repeat', repeatModeToCapability[this.repeat] ?? 'none');
            }
            if (this.#device.hasCapability('speaker_shuffle')) {
                await this.#device.setCapabilityValue('speaker_shuffle', this.shuffle);
            }

            this.#scheduleNowPlayingAppUpdate(
                client.isPlaying ? client.bundleIdentifier : null,
                client.isPlaying ? client.displayName : null
            );

            await this.#setArtwork(client);
            this.#emitMiniPlayerUpdate();
        } catch (err) {
            this.log(this.deviceName, 'Failed to update now playing info', err);
        }
    }

    async #syncDynamicCapabilities(client: AirPlayClient): Promise<void> {
        const capabilities: Array<[string, Proto.Command]> = [
            ['speaker_next', Proto.Command.NextTrack],
            ['speaker_prev', Proto.Command.PreviousTrack],
            ['speaker_repeat', Proto.Command.ChangeRepeatMode],
            ['speaker_shuffle', Proto.Command.ChangeShuffleMode]
        ];

        for (const [id, command] of capabilities) {
            const isSupported = client.isCommandSupported(command);
            const hasCapability = this.#device.hasCapability(id);

            if (isSupported && !hasCapability) {
                await this.#device.addCapability(id);
            } else if (!isSupported && hasCapability) {
                await this.#device.removeCapability(id);
            }
        }

        // Volume set is dynamically managed for Apple TV based on output device capabilities.
        if (this.#device instanceof AppleTVDevice) {
            const isVolumeAvailable = this.#sdkDevice?.state.volumeAvailable ?? false;
            const hasVolumeSet = this.#device.hasCapability('volume_set');

            if (isVolumeAvailable && !hasVolumeSet) {
                await this.#device.addCapability('volume_set');
            } else if (!isVolumeAvailable && hasVolumeSet) {
                await this.#device.removeCapability('volume_set');
            }
        }
    }

    #getFeatureAvailability(): { previous: boolean; next: boolean; shuffle: boolean; repeat: boolean } {
        const state = this.#sdkDevice?.state;

        if (!state?.activeClient) {
            return {previous: false, next: false, shuffle: false, repeat: false};
        }

        return {
            previous: state.isCommandSupported(Proto.Command.PreviousTrack),
            next: state.isCommandSupported(Proto.Command.NextTrack),
            shuffle: state.isCommandSupported(Proto.Command.ChangeShuffleMode),
            repeat: state.isCommandSupported(Proto.Command.ChangeRepeatMode)
        };
    }

    #emitMiniPlayerUpdate(): void {
        clearTimeout(this.#miniPlayerUpdateTimer);
        this.#miniPlayerUpdateTimer = setTimeout(() => {
            this.#device.homey.api.realtime('apple-mini-player-update', this.getState())
                .catch((err: unknown) => this.log(this.deviceName, 'Failed to emit mini player update:', err));
        }, 100);
    }

    async #serialized(fn: () => Promise<void>): Promise<void> {
        let releaseLock: () => void;
        const newLock = new Promise<void>(resolve => {
            releaseLock = resolve;
        });

        const previousLock = this.#updateLock;
        this.#updateLock = newLock;

        await previousLock;

        try {
            await fn();
        } finally {
            releaseLock!();
        }
    }

    #scheduleNowPlayingAppUpdate(bundleIdentifier: string | null, displayName: string | null): void {
        clearTimeout(this.#nowPlayingAppTimer);
        this.#nowPlayingAppTimer = setTimeout(() => {
            this.#updateNowPlayingAppImpl(bundleIdentifier, displayName)
                .catch(err => this.log(this.deviceName, 'Failed to update now playing app', err));
        }, 1000);
    }

    async #updateNowPlayingAppImpl(bundleIdentifier: string | null, displayName: string | null): Promise<void> {
        if (!this.#device.hasCapability('now_playing_app')) {
            return;
        }

        const currentNowPlayingApp = this.#device.getCapabilityValue('now_playing_app');

        if (currentNowPlayingApp === displayName) {
            return;
        }

        await this.#device.setCapabilityValue('now_playing_app', displayName);

        if (this.#device instanceof AppleTVDevice) {
            await this.app.appleTvFlow.triggerNowPlayingAppChanges(this.#device, bundleIdentifier ?? '-', displayName ?? '-');
        }
    }
}

File: src/logic/airplay.py
from __future__ import annotations

import asyncio
import inspect
import time
from typing import TYPE_CHECKING, Any

import pyatv.interface as pyatv_interface
from pyatv.const import DeviceState, PowerState, RepeatState, ShuffleState

if TYPE_CHECKING:
    from homey.device import Device
    from homey.image import Image

    from ..app import AppleApp

_DEBOUNCE_DELAY_S = 0.3


class AirPlayLogic(pyatv_interface.PushListener, pyatv_interface.PowerListener):
    """
    Receives push updates (now-playing state and power state) from pyatv and
    syncs them to Homey capabilities.

    Shared between Apple TV and HomePod devices.
    """

    @property
    def device_name(self) -> str:
        return self._device.get_name()

    def __init__(self, device: Device, app: AppleApp) -> None:
        self._device = device
        self._app = app
        self._artwork: Image | None = None
        self._artwork_hash: str | None = None
        self._debounce_task: asyncio.Task | None = None
        self._update_task: asyncio.Task | None = None
        self._pending_playing: pyatv_interface.Playing | None = None
        self._atv: pyatv_interface.AppleTV | None = None
        self._shuffle: bool = False
        self._repeat: str = 'off'
        self._position_update_time: float = 0.0

    # ------------------------------------------------------------------
    # Lifecycle
    # ------------------------------------------------------------------

    async def _call_image_method(self, method_name: str, *args: Any) -> None:
        """Call a method on the artwork image, awaiting if it returns a coroutine."""
        if self._artwork is None:
            return
        method = getattr(self._artwork, method_name, None)
        if callable(method):
            result = method(*args)
            if inspect.isawaitable(result):
                await result

    async def initialize(self) -> None:
        """Register artwork image with Homey and clear now-playing state."""
        self._artwork = await self._device.homey.images.create_image()
        try:
            await self._device.set_album_art_image(self._artwork)
            await self.clear_now_playing()
            await self.update_artwork_url()
        except Exception:
            await self._device.homey.images.unregister_image(self._artwork)
            self._artwork = None
            raise

    async def uninitialize(self) -> None:
        """Stop push updates and clean up the artwork image."""
        self.stop()
        if self._artwork is not None:
            try:
                await self._device.homey.images.unregister_image(self._artwork)
            except Exception:
                pass
            self._artwork = None

    def set_atv(self, atv: pyatv_interface.AppleTV) -> None:
        """Attach pyatv interface and register as push + power listener."""
        self._atv = atv

        atv.push_updater.listener = self
        atv.push_updater.start(initial_delay=0)

        # Power listener is only available on devices with Companion protocol.
        try:
            atv.power.listener = self
        except Exception:
            pass

    def stop(self) -> None:
        """Stop push updates (called on disconnect / uninit)."""
        if self._debounce_task is not None:
            self._debounce_task.cancel()
            self._debounce_task = None

        if self._update_task is not None and not self._update_task.done():
            self._update_task.cancel()
            self._update_task = None

        if self._atv is not None:
            try:
                self._atv.push_updater.stop()
            except Exception:
                pass
            try:
                self._atv.power.listener = None
            except Exception:
                pass

    # ------------------------------------------------------------------
    # PushListener implementation
    # ------------------------------------------------------------------

    def playstatus_update(
        self,
        updater: pyatv_interface.PushUpdater,
        playing: pyatv_interface.Playing,
    ) -> None:
        asyncio.create_task(self._on_playing_update(playing))

    def playstatus_error(
        self,
        updater: pyatv_interface.PushUpdater,
        exception: Exception,
    ) -> None:
        self._device.error('Push update error:', exception)

    # ------------------------------------------------------------------
    # PowerListener implementation
    # ------------------------------------------------------------------

    def powerstate_update(
        self,
        old_state: PowerState,
        new_state: PowerState,
    ) -> None:
        asyncio.create_task(self._handle_power_state(new_state))

    # ------------------------------------------------------------------
    # Now-playing state
    # ------------------------------------------------------------------

    async def clear_now_playing(self) -> None:
        """Reset all now-playing capabilities to their default/empty values."""
        try:
            self._artwork_hash = None

            await self._call_image_method('set_url', '')
            await self._call_image_method('update')

            await self._update_now_playing_app(None, None)

            await self._device.set_capability_value('speaker_album', '')
            await self._device.set_capability_value('speaker_artist', '')
            await self._device.set_capability_value('speaker_track', '')
            await self._device.set_capability_value('speaker_duration', -1)
            await self._device.set_capability_value('speaker_position', -1)
            await self._device.set_capability_value('speaker_playing', False)

            self._shuffle = False
            self._repeat = 'off'

            self._device.log(self.device_name, 'Now playing info cleared.')
        except Exception as err:
            self._device.log(self.device_name, 'Failed to clear now playing info', err)

    async def update_artwork_url(self) -> None:
        """Push the current artwork cloud/local URLs to the capability."""
        if not self._device.has_capability('artwork_url'):
            return

        if self._artwork is None:
            return

        cloud_url = getattr(self._artwork, 'cloud_url', None)
        local_url = getattr(self._artwork, 'local_url', None)

        base_url = cloud_url or local_url
        if not base_url:
            return

        cache_buster = int(time.time() * 1000)
        artwork_url = f'{base_url}?v={cache_buster}'
        await self._device.set_capability_value('artwork_url', artwork_url)
        self._device.log(self.device_name, 'Artwork URL updated.', artwork_url)

        cloud_url_with_cache = f'{cloud_url}?v={cache_buster}' if cloud_url else ''
        local_url_with_cache = f'{local_url}?v={cache_buster}' if local_url else ''

        if self._device.has_capability('artwork_url_local'):
            await self._device.set_capability_value('artwork_url_local', local_url_with_cache)

        if self._device.has_capability('artwork_url_cloud'):
            await self._device.set_capability_value('artwork_url_cloud', cloud_url_with_cache)

        await self._trigger_artwork_url_updated(local_url_with_cache, cloud_url_with_cache)
        await self._emit_mini_player_update()

    async def _trigger_artwork_url_updated(self, local_url: str, cloud_url: str) -> None:
        from ..apple_tv.device import AppleTVDevice
        from ..homepod_base.device import HomePodBaseDevice

        if isinstance(self._device, AppleTVDevice):
            await self._app.apple_tv_flow.trigger_artwork_url_updated(
                self._device, local_url, cloud_url
            )
        elif isinstance(self._device, HomePodBaseDevice):
            await self._app.homepod_flow.trigger_artwork_url_updated(
                self._device, local_url, cloud_url
            )

    # ------------------------------------------------------------------
    # Power state
    # ------------------------------------------------------------------

    async def _handle_power_state(self, new_state: PowerState) -> None:
        """Handle power state change events."""
        self._device.log('Power state changed:', new_state)

        is_on = new_state == PowerState.On

        try:
            if self._device.has_capability('onoff'):
                await self._device.set_capability_value('onoff', is_on)

            if self._device.has_capability('power'):
                await self._device.set_capability_value(
                    'power',
                    self._device.homey.i18n.translate(
                        'capability.power.on' if is_on else 'capability.power.off'
                    ),
                )
        except Exception as err:
            self._device.error('Failed to set power state:', err)

        if not is_on:
            await self.clear_now_playing()

    # ------------------------------------------------------------------
    # Debounced now-playing updates
    # ------------------------------------------------------------------

    async def _on_playing_update(self, playing: pyatv_interface.Playing) -> None:
        """Debounced handler for push-update playback state changes."""
        if playing.position is not None:
            self._position_update_time = time.time()

        self._pending_playing = playing

        if self._debounce_task is not None:
            self._debounce_task.cancel()

        self._debounce_task = asyncio.create_task(self._debounced_update())

    async def _debounced_update(self) -> None:
        try:
            await asyncio.sleep(_DEBOUNCE_DELAY_S)
        except asyncio.CancelledError:
            return

        playing = self._pending_playing
        if playing is not None:
            self._update_task = asyncio.create_task(self._update_now_playing(playing))

    async def _update_now_playing(self, playing: pyatv_interface.Playing) -> None:
        """Update Homey capabilities from a Playing object."""
        if (
            self._device.has_capability('onoff')
            and self._device.get_capability_value('onoff') is False
        ):
            return

        device_state = playing.device_state
        is_playing = device_state == DeviceState.Playing

        self._device.log(
            self.device_name, 'Now playing update',
            f'state={device_state}',
            f'title={playing.title!r}',
            f'artist={playing.artist!r}',
            f'album={playing.album!r}',
            f'hash={playing.hash!r}',
        )

        try:
            await self._device.set_capability_value('speaker_playing', is_playing)

            if playing.album is not None:
                await self._device.set_capability_value('speaker_album', playing.album)

            if playing.title is not None:
                await self._device.set_capability_value('speaker_track', playing.title)

            if playing.total_time is not None:
                await self._device.set_capability_value('speaker_duration', playing.total_time)

            if playing.position is not None:
                await self._device.set_capability_value('speaker_position', playing.position)

            # Update volume_set capability from the audio interface.
            if self._device.has_capability('volume_set') and self._atv is not None:
                try:
                    volume = self._atv.audio.volume
                    if volume is not None:
                        await self._device.set_capability_value('volume_set', volume / 100.0)
                except Exception:
                    pass

            # Shuffle / repeat state
            shuffle_state = getattr(playing, 'shuffle', None)
            if shuffle_state is not None:
                self._shuffle = shuffle_state != ShuffleState.Off

            repeat_state = getattr(playing, 'repeat', None)
            if repeat_state == RepeatState.Track:
                self._repeat = 'one'
            elif repeat_state == RepeatState.All:
                self._repeat = 'all'
            elif repeat_state is not None:
                self._repeat = 'off'

            # Artwork
            artwork_hash = playing.hash
            if artwork_hash != self._artwork_hash:
                await self._fetch_artwork(artwork_hash)

            # Now-playing app — resolve before artist fallback.
            app_name: str | None = None
            if is_playing and self._atv is not None:
                try:
                    app = self._atv.metadata.app
                except Exception:
                    app = None

                if app is not None:
                    app_name = getattr(app, 'name', None)
                    await self._update_now_playing_app(
                        getattr(app, 'identifier', None),
                        app_name,
                    )
                else:
                    await self._update_now_playing_app(None, None)
            else:
                await self._update_now_playing_app(None, None)

            # Artist — fall back to app name when artist is not available
            # (e.g. video content on Netflix, YouTube, etc.).
            artist = playing.artist if playing.artist is not None else app_name
            if artist is not None:
                await self._device.set_capability_value('speaker_artist', artist)

            await self._emit_mini_player_update()

        except Exception as err:
            self._device.log(self.device_name, 'Failed to update now playing info', err)

    async def _fetch_artwork(self, artwork_hash: str | None) -> None:
        """Fetch and update artwork from pyatv."""
        if self._atv is None or self._artwork is None:
            return

        try:
            artwork_info = await self._atv.metadata.artwork()

            if artwork_info is None or artwork_info.bytes is None:
                await self._update_artwork_data(None)
                return

            # Skip HEIC artwork — Homey cannot display it.
            mimetype = getattr(artwork_info, 'mimetype', '') or ''
            if 'heic' in mimetype.lower() or 'heif' in mimetype.lower():
                self._device.log(
                    self.device_name,
                    f'Skipping HEIC artwork (mimetype={mimetype})',
                )
                return

            await self._update_artwork_data(artwork_info.bytes)
            self._artwork_hash = artwork_hash
        except Exception as err:
            self._device.error(self.device_name, 'Failed to fetch artwork:', err)

    async def _update_artwork_data(self, data: bytes | None) -> None:
        """Push raw artwork bytes (or None) into the Homey Image."""
        if self._artwork is None:
            return

        try:
            if data:
                async def write_to_stream(stream: Any) -> None:
                    stream.write(data)

                await self._call_image_method('set_stream', write_to_stream)

            await self._call_image_method('update')
            await self.update_artwork_url()
        except Exception as err:
            self._device.error(self.device_name, 'Failed to update artwork data:', err)

    async def _emit_mini_player_update(self) -> None:
        """Emit a realtime event so mini player widgets can update without polling."""
        def cap(name: str) -> Any:
            try:
                return self._device.get_capability_value(name)
            except Exception:
                return None

        try:
            await self._device.homey.api.realtime('apple-mini-player-update', {
                'deviceId': self._device._id,
                'deviceName': self._device.get_name(),
                'track': cap('speaker_track'),
                'artist': cap('speaker_artist'),
                'album': cap('speaker_album'),
                'playing': cap('speaker_playing'),
                'position': cap('speaker_position'),
                'duration': cap('speaker_duration'),
                'volume': cap('volume_set'),
                'artworkUrl': cap('artwork_url'),
                'shuffle': self._shuffle,
                'repeat': self._repeat,
                'positionTimestamp': int(self._position_update_time * 1000),
            })
        except Exception as err:
            self._device.error(self.device_name, 'Failed to emit mini player update:', err)

    async def _update_now_playing_app(
        self,
        bundle_identifier: str | None,
        display_name: str | None,
    ) -> None:
        if not self._device.has_capability('now_playing_app'):
            return

        current = self._device.get_capability_value('now_playing_app')
        if current == display_name:
            return

        self._device.log(
            self.device_name, 'Now playing app changed.', bundle_identifier, display_name
        )

        await self._device.set_capability_value('now_playing_app', display_name)

        from ..apple_tv.device import AppleTVDevice
        if isinstance(self._device, AppleTVDevice):
            await self._app.apple_tv_flow.trigger_now_playing_app_changes(
                self._device,
                bundle_identifier or '-',
                display_name or '-',
            )
